homeassistant/components/tradfri/coordinator.py

```diff
@@ -51,27 +51,23 @@ async def set_hub_available(self, available: bool) -> None:
     @callback
     def _observe_update(self, device: Device) -> None:
         """Update the coordinator for a device when a change is detected."""
-        self.update_interval = timedelta(seconds=SCAN_INTERVAL)  # Reset update interval
-
         self.async_set_updated_data(data=device)
 
     @callback
-    def _exception_callback(self, device: Device, exc: Exception | None = None) -> None:
+    def _exception_callback(self, exc: Exception) -> None:
         """Schedule handling exception.."""
-        self.hass.async_create_task(self._handle_exception(device=device, exc=exc))
+        self.hass.async_create_task(self._handle_exception(exc))
 
-    async def _handle_exception(
-        self, device: Device, exc: Exception | None = None
-    ) -> None:
+    async def _handle_exception(self, exc: Exception) -> None:
         """Handle observe exceptions in a coroutine."""
-        self._exception = (
-            exc  # Store exception so that it gets raised in _async_update_data
-        )
+        # Store exception so that it gets raised in _async_update_data
+        self._exception = exc
 
-        _LOGGER.debug("Observation failed for %s, trying again", device, exc_info=exc)
-        self.update_interval = timedelta(
-            seconds=5
-        )  # Change interval so we get a swift refresh
+        _LOGGER.debug(
+            "Observation failed for %s, trying again", self.device, exc_info=exc
+        )
+        # Change interval so we get a swift refresh
+        self.update_interval = timedelta(seconds=5)
         await self.async_request_refresh()
 
     async def _async_update_data(self) -> Device:
```

Comment:
I'm a bit confused by this coordinator as it's using the coordinator polling mechanism to raise exceptions and as a health check to make sure the observation connection is established. I see how it's made to work, but it feels dirty.

Reply:
Ok. I'll like to do that in a follow up as this PR is only fixing the current implementation.

Plan:
1. Merge this PR.
2. Add complete test coverage for observations.
3. Rewrite observation coordinator.